Documentation/StringFormatting.md

```diff
@@ -0,0 +1,128 @@
+# String Formatting
+
+Many places in Serenity allow you to format strings, similar to `printf()`, for example `String::formatted()`, `StringBuilder::appendff()`, or `dbgln()`. These are checked at compile time to ensure the format string matches the number of parameters. The syntax is largely based on the [C++ `std::formatter` syntax](https://en.cppreference.com/w/cpp/utility/format/formatter#Standard_format_specification) but there are some differences.
+
+For basic usage, any occurrences of `{}` in the format string are replaced with the other arguments, converted to string form, in order:
+
+```c++
+String::formatted("Well, {} my {} friends!", "hello", 42) == "Well, hello my 42 friends!";
+```
+
+If you want to include a literal `{` in the output, use `{{`:
+
+```c++
+String::formatted("{{ {}", "hello") == "{ hello";
+```
+
+You can refer to the arguments by index, if you want to repeat one or change the order:
+```c++
+String::formatted("{3}{0}{1}", "a", "b", "c", "d") == "dab";
```

Comment:
This actually won't compile, as the `"c"` argument is not used in the format string.

Reply:
(though the runtime is perfectly okay with evaluating this)

Reply:
Thanks. Seems weird to me that `TestFormat.cpp` has test cases making sure things work that are bypassing the compile-time check so can't actually be used. If they're valid, the check should allow them; if they're not, why are we running tests to make sure they work?